---
title: "Generating Heatmaps in R"
author: "Fong Chun Chan"
date: "August 25, 2015"
output: html_document
---

## Objectives 

* Learn what heatmaps are
* Learn how to generate a heatmap in R
* Learn the importance of scaling and clustering in heatmap generation

## What are Heatmaps?

Let's say you were presented with the following 5x5 table/matrix. 

```{r, message = FALSE}
library("dplyr")
library("knitr")
set.seed(1)

mat <- matrix(rnorm(25), 5, 5, 
              dimnames = list(LETTERS[1:5], 1:5))
mat %>%
  as.data.frame %>%
  kable
```

Your first intuition may be whether there is anything interesting in this matrix/table?  It's not immediately obvious which values in the matrix may be outliers/extremes. But if we convert this matrix into a graphical representation:

```{r}
# mat[5:1] is done to make the heatmap output in the same order as the table 
# above.
heatmap(mat[5:1, ], Rowv = NA, Colv = NA, scale = "none")
```

Each cell in the matrix has now been converted into a different color. Now you can immediately see the red cells in the heatmap which indicate that (D,5) and (D,3) appear to be outliers in this data. Referring back to the matrix confirms this as `r mat["D", 5]` (D,5) and `r mat["D", 3]` (D,3) are the most extreme negative values in the matrix. 

This example demonstrates the utility of heatmaps which are graphical representations of individual values by different colors. You can find examples of heatmaps in various fields ([Wikipedia has numerous example](https://en.wikipedia.org/wiki/Heat_map#Examples)). 

> It's important to note that a heatmap does not have to be in a "matrix/table" form

For instance, [Wikipedia has an example of a "Choropleth heatmap"](https://commons.wikimedia.org/wiki/File%3A2012_US_Senate_election_in_Massachusetts_results_by_municipality.svg).

![Wikipedia Example of Choropleth Heatmap](https://upload.wikimedia.org/wikipedia/commons/2/28/2012_US_Senate_election_in_Massachusetts_results_by_municipality.svg "Wikipedia Example of Choropleth Heatmap")

Here the different regions on the map are represented with different colors to indicate how the votes are distributed between the two candidates. **This is also considered a heatmap** and provides a nice graphical representation of which candidate is doing better in certain regions. Without a heatmap to represent such data, it would be difficult to see such patterns.

## Generating Heatmaps from Matrix Data in R

Now that you know what heatmaps are, let's see how we can generate these heatmaps in R when the data is in matrix form. Let's use the same matrix `mat` from the beginning. 

```{r}
set.seed(1)
mat <- matrix(rnorm(25), 5, 5, 
              dimnames = list(LETTERS[1:5], 1:5))

mat
```

The easiest way to generate a heatmap in R is to use the `image` function:

```{r}
image(mat)
```

You might notice how this looks a bit different. Specifically the order of the cells appears to be different. As it turns out, the `image` function transpose the data such that the rows become columns and columns become rows. We can fix this by using the `t` (transpose) function on the matrix before using the `image` function

```{r }
image(t(mat))
```

This is better now except the order of the rows seems to be different from the input matrix. Specifically, the `image` function actually flips the ordering of the rows. We can fix this:

```{r}
image(t(mat[nrow(mat):1, ]))
```

All we are doing here is reversing the rows before we pass it into the `image` function. Such that when the function reverses it again it will be in the same order as the input (perhaps there is a better solution to this). 

This is almost what we want except we have lost all our x and y labels! There is a way around this where you can manually add the labels, but at this point it's better to just introduce the `heatmap` function. The `heatmap` function is available from base R. We can use it like this:

```{r}
# mat[5:1] is done to make the heatmap output in the same order as the table 
# above.
heatmap(mat[5:1, ], Rowv = NA, Colv = NA, scale = "none")
```

This is in fact the same code that I used at the beginning of the post. The `heatmap` function actually calls `t(mat)` to generate the heatmap and thus it will result in the same row re-ordering issue we mentioned already. Hence, why the row re-ordering code is used again. The benefit of this is that we also get the x- and y-labels added automatically for us.  **You may have noticed additional parameters that I have set (Rowv, Colv, scale). I will explain how to use these in the next section**.

If you want to change to use a different color scheme. The `col` parameter allows you (actually belongs to the `image` function) modify this. By default the color scheme is set to `heat.colors(12)`:

```{r}
heat.colors(12)
```

This basically gives us a color palette that can be correlated to values in the matrix. The `RColorBrewer` R package provides a way to make your color palette:

```{r}
library("RColorBrewer")
my.palette <- colorRampPalette(c("green", "black", "red"))(n = 12)
heatmap(mat[5:1, ], Rowv = NA, Colv = NA, scale = "none", col = my.palette)
```

Here we setting a color gradient such as it goes from green -> black -> red which corresponds to low -> mid -> high values. You can play around this to get the exact color scheme you want. In addition to the base `heatmap` function, there have been a handful of R packages that have improved on the base `heatmap` function (e.g. [gplots - heatmap.2](https://cran.r-project.org/web/packages/gplots/index.html), [heatmap.plus](https://cran.r-project.org/web/packages/heatmap.plus/index.html), [d3heatmap](https://cran.r-project.org/web/packages/d3heatmap/index.html)). The newest one (to the best of my knowledge and at the time of this writing) is d3heatmap from RStudio. This package leverages off [D3](http://d3js.org/) to provide an interactive heatmap:

```{r}
library("d3heatmap")
d3heatmap(mat, Rowv = NA, Colv = NA, colors = heat.colors(12))
```

You can now hover over the cells and see the individual values, zoom in and out. The function also no longer does the row re-ordering so we don't need to do that hack to maintain the initial ordering. 

## Heatmap Normalization

While heatmaps provide a fantastic way to graphical visualize values, but they may not always automatically work right out of the box. I will use an example from the biological sciences to illustrate this. 

Heatmaps are often used in this field to represent "gene expression" matrices. Specifically, the rows represent different biological features (genes) and columns represent different samples. Each cell in the matrix then represents the "expression" of that specific biological feature in a specific sample. Below is an example of one of these matrices and an attempt to generate an heatmap on it:

> This matrix is from [Lim et al. Comprehensive miRNA sequence analysis reveals survival differences in diffuse large B-cell lymphoma patients. Genome Biology](http://www.genomebiology.com/2015/16/1/18). 2015.

```{r}
mat.file <- "data/mirna-normalized-expression-subset-lim-2015.txt"
mat.lim <- as.matrix(read.table(mat.file, sep = "\t"))

d3heatmap(mat.lim, colors = rev(brewer.pal(9, "RdBu")), Rowv = NA, Colv = NA,
          width = 750, height = 1000, yaxis_width = 180)
```

That looks really funny hey? What is actually happening? We see a strong "striping" pattern occurring across the rows. To gain some insight into the issue, we can take a look at distribution of expression feature we are working with.

```{r message = FALSE, fig.width = 10}
library("dplyr")
library("reshape2")
library("ggplot2")

mat.lim %>%
	melt %>%  
	ggplot(aes(x = Var1, y = value)) +
	geom_boxplot() +
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
```

The boxplot above shows the distribution of expression values for each feature across the samples. Immediately, certain features are much more highly expressed in all samples compared to other features. And similarly some features are much more lowly expressed than all other features. What's basically happening is that the **between-feature differences** are much greater than the between sample differences. A highly expressed feature will appear a row of all the same colors. For instance, 

```{r}
mira.highest <- mat.lim %>%
	melt %>%
	group_by(Var1) %>%
	summarize(mean.vals = mean(value)) %>%
	arrange(desc(mean.vals)) %>%
	slice(1) %>%
	.$Var1 %>%
	as.character
```

The `r mira.highest` is the most highly expressed feature in the dataset. Can you identify which row in the above heatmap it is (red means high expression)?  It should be one of the rows that is basically almost all red. So how do we get around this? This is where scaling comes in. 

What we really want to do is to put every row onto the same scale. Right now the range of each row is a function of the features expression level. What we can do is perform what is called Z-score normalization. Specifically:

$$ z = \frac{x - \mu}{\sigma} $$

Where each value in a row, $x$, is subtracted by the row mean, $\mu$, and then divided by standard deviation $\sigma$. This effectively converts each value into a z-score and places all rows onto the same scale. The `d3heatmap` function provides the ability to do a z-score row normalization by using the `scale = "row"`:

```{r}
d3heatmap(mat.lim, colors = rev(brewer.pal(9, "RdBu")), Rowv = NA, Colv = NA,
          width = 750, height = 1000, yaxis_width = 180, scale = "row")
```

This looks a bit better. We no longer have the "stripe" patterns in the rows anymore. 

## Heatmap Clustering

We've improved things a bit, but now what we see are columns/samples that have similar patterns. It would be interested to actually group these samples together. This is where the concept of clustering comes in. Clustering is a machine learning technique that aims to "group" together (i.e. cluster) similar things. 

The `d3heatmap` function (and `heatmap`) actually performs the clustering automatically for you for both rows and columns. In the previous section, I had turned it off using `Rowv = NA` and `Colv = NA` to demonstrate only the heatmap component of the function. Let's see what happens if we enable clustering following normalization.

```{r}
d3heatmap(mat.lim, scale = "row", Rowv = NA, colors = rev(brewer.pal(9, "RdBu")), 
          width = 750, height = 1000, yaxis_width = 180)
```

This is a nice final output that you often see in biological papers. We can clearly see two "clusters" as indicated by the dendrogram/tree on the top. The samples on the left cluster demonstrates higher expression (red) while the samples in the right cluster exhibit lower expression for these features (blue). In fact, this actually makes perfect sense because the samples on the side of the dendrogram is a completely different set of biological samples from the right. 

## Conclusions

I hope this tutorial has helped you understand what heatmaps are. In addition to generating heatmaps, you should also understand the importance of normalization and clustering of data to produce the final outputs you see in biological science papers. This is just the tip of the iceberg though. Some further areas of interest in generating heatmaps may be:

* [Exploring different forms of clustering](http://stackoverflow.com/questions/6806762/setting-distance-matrix-and-clustering-methods-in-heatmap-2)
* [Extracting the dendrograms](http://seqanswers.com/forums/showthread.php?t=32279)
* [Adding annotation tracks for rows and samples](https://www.biostars.org/p/18211/)

## Contact

Feel free to contact me for help regarding the content in this workshop:

* email: fongchunchan@gmail.com
* twitter: [https://twitter.com/fongchunchan](https://twitter.com/fongchunchan)
* blog: [http://tinyheero.github.io/](http://tinyheero.github.io/)
